services/s3/src/test/java/software/amazon/awssdk/services/s3/S3EndpointResolutionTest.java

```diff
@@ -508,4 +512,74 @@ private S3Configuration withDualstackAndAccelerateEnabled() {
+                .hasMessageContaining(USE_ARN_REGION_MESSAGE);
+    }
+
+    @Test
+    public void accessPointArn_us_east_1_region_s3_external_1_useArnRegionFalse_throwsIllegalArgumentException() throws Exception {
+        mockHttpClient.stubNextResponse(mockListObjectsResponse());
+        S3Client s3Client = clientBuilder().region(Region.of("s3-external-1")).build();
+        String accessPointArn = "arn:aws:s3:us-east-1:12345678910:accesspoint:foobar";
+        assertThatThrownBy(() -> s3Client.listObjects(ListObjectsRequest.builder().bucket(accessPointArn).build()))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining(USE_ARN_REGION_MESSAGE);
+    }
+
+    @Test
+    public void accessPointArn_us_gov_west_1_clientRegion_fips_us_gov_east_1_useArnRegionTrue() throws Exception {
+        mockHttpClient.stubNextResponse(mockListObjectsResponse());
+        S3Client s3Client = clientBuilder().region(Region.of("fips-us-gov-east-1"))
+                .serviceConfiguration(S3Configuration.builder().useArnRegionEnabled(true).build()).build();
+        String accessPointArn = "arn:aws-us-gov:s3:us-gov-west-1:123456789012:accesspoint:myendpoint";
+
+        assertThatThrownBy(() -> s3Client.getObject(GetObjectRequest.builder().bucket(accessPointArn).key("someKey").build()))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining(CROSS_REGION_FIPS_ERROR);
+    }
+
+    @Test
+    public void accessPointArn_us_gov_west_1_clientRegion_fips_us_gov_east_1_useArnRegionFalse() throws Exception {
+        mockHttpClient.stubNextResponse(mockListObjectsResponse());
+        S3Client s3Client = clientBuilder().region(Region.of("fips-us-gov-east-1"))
+                .serviceConfiguration(S3Configuration.builder().useArnRegionEnabled(false).build()).build();
+        String accessPointArn = "arn:aws-us-gov:s3:us-gov-west-1:123456789012:accesspoint:myendpoint";
+
+        assertThatThrownBy(() -> s3Client.getObject(GetObjectRequest.builder().bucket(accessPointArn).key("someKey").build()))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining(CROSS_REGION_FIPS_ERROR);
+    }
+
+    @Test
+    public void accessPointArnRegion_fips_clientRegion_non_fips_useArnRegionTrue() throws Exception {
+        mockHttpClient.stubNextResponse(mockListObjectsResponse());
+        S3Client s3Client = clientBuilder().region(Region.of("us-west-1"))
+                .serviceConfiguration(S3Configuration.builder().useArnRegionEnabled(true).build()).build();
+        String accessPointArn = "arn:aws-us-gov:s3:fips-us-gov-east-1:123456789012:accesspoint:myendpoint";
+
+        assertThatThrownBy(() -> s3Client.getObject(GetObjectRequest.builder().bucket(accessPointArn).key("someKey").build()))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining(CROSS_REGION_FIPS_ERROR);
+    }
```

Comment:
Can we add a test similar to this one, but with a client region of us-gov-east-1 (and an ARN region of fips-us-gov-east-1)? The customer would probably expect this call to go to the fips endpoint, so it might break compliance for them if it goes to a non-fips endpoint.

I'd honestly vote for throwing an exception in this case, but I'd be okay with it going to the FIPS endpoint as well. Either way, we should align that behavior with the other SDKs.

Reply:
The scenario mentioned by you is same as Scenario5 as mentioned in above table in Motivation and Context section.
Current V2 code does not gets the partition correctly for Fips region/Gov regions and assigns them as default partition "aws'. And when we do a partition check which is right below Region match check it fails.

Reply:
I'm suggesting the inverse of scenario 5, swapping the client and arn region columns. It might be more interesting if it's in the AWS partition, so something like fips-us-west-2 for the ARN region and us-west-2 for the client region.

Reply:
Got it . It successfully resolved endpoints in this case. Added  test cases for this.

Reply:
For ARN region of fips-us-gov-east , now we throw exception ```Invalid ARN, FIPS region not allowed in ARN```